### PR TITLE
Fix #457: Load delay during gamelist processing

### LINKED ISSUE #457: Slow initial startup show ES title screen for >3 minutes with no progress bar.
I have many thousands of roms present, some of them without game list files. In this situation, ES takes forever to load on power up (~3 minutes). I think it is looping over all the roms to find the filenames. I don't understand why this is so slow, since it is not scraping the roms.

It would be nice to a) speed up the launch of ES on boot, b) display some kind of progress or at least a number in the bottom left corner, such as "6,125 roms found" with a number that increases.

### PR DESCRIPTION
**Pull request by aarononeal**
The problem causing slow loading is because of how relative paths are handled. Both the game loader and the theme loader rely on the following functions: `resolveRelativePath`, `createRelativePath`, `removeCommonPath`.

These methods use `isDirectory()` to determine whether a file should search parent directories when referenced by `.` instead of the one it is in. This is relevant for themes which load parent relative content.

This is not relevant for loading games and so every single file of every game list hits the file system multiple times to see if a path is a directory. If you're loading from the network or slow media this can take several minutes while every game is probed.

The fix adds `_searchParents` so that the theme system does (true) and the game loader does not (false).

You also need to set the `ParseGamelistOnly` setting to avoid touching the file system on load.

These changes bring an 8 minute load time down to about 3 seconds.

Added a comment as a warning to hopefully avoid re-introduction of `isDirectory()` in that path in the future.

Please ensure I have not overlooked a game loading use case that actually requires the parent path flow.

### CODE REVIEW

**Comment by cmitu** [comment]
> This is not relevant for loading games and so every single file of every game list hits the file system multiple times to see if a path is a directory. If you're loading from the network or slow media this can take several minutes while every game is probed.

Why wouldn't it be relevant ? The `path` in a gamelist can be relative, so EmulationStation should know if the file is still present or not, to display it.

**Comment by aarononeal** [comment]
> Why wouldn't it be relevant ? The `path` in a gamelist can be relative ...

The relative path is still resolved. 

It's just that the gamelist case has no need to traverse up to a parent that I can tell. If you pass in `./games` the call to `isDirectory()` will succeed if the folder exists, and if it doesn't, there is no reason to assume that traversing up to the parent `.` is the correct behavior.

This directory/parent logic seems to be needed more by the theme system. I didn't trace down all the paths to see why, just left them untouched.

> ... EmulationStation should know if the file is still present or not, to display it.

When `ParseGamelistOnly` is set, ES should not be checking for the presence of entries or touching the filesystem. Checking is extremely slow, and redundant for when lists are actively managed, which is why the setting exists.

Even when not set, best to avoid hitting the filesystem as much as possible. It's unclear why `isDirectory()` is needed for the game load codepath to have the correct behavior. And clearly it has significantly negative impact to load times (160x slower).

**Comment by pjft** [comment]
Thanks for putting this together. I did test it out on a network drive, and it worked correctly, but I didn't experience any change in performance on my end - was already on ParseGamelistsOnly.

I'd like to get some people to test this out a bit more, especially around the edge cases you and @cmitu were discussing. Any chance you could share this in the forums for testing from some other folks with different setups, just to make sure that there's no edge case or any other issue here?

Thanks.

**Comment by joolswills** [comment]
Needs to be checked with symlinks also etc.

**Comment by rsn8887** [comment]
My issue https://github.com/RetroPie/EmulationStation/issues/457 pertains to parse gamelists only being OFF. Does this fix speed up that case even?

**Comment by aarononeal** [comment]
No. However, this dramatically improves I/O and load times in the case that game lists are trusted.

To speed up the full scan case, a similar set of optimizations are required to a different codepath to reduce `stat` calls as was done here.

**Comment by aarononeal** [comment]
Will this be merged?

**Comment by tomaz82** [comment]
I don't understand the need for this PR, or even the describing text, the purpose of the call to isDirectory() is not to search a parent, it's to make sure that relativeTo doesn't contain a file name.

**Comment by aarononeal** [comment]
That is not necessary in the case that the game list is trusted which is the path this PR optimizes. Hitting the file system is detrimental to performance over network drives in this case.

**Comment by tomaz82** [comment]
Still, the ext here suggests that the functions are used to search parents, which it isn't ( that's just a sideeffect that should probably never happen ). The purpose of the function is to make sure that relateiveTo does not contain a file name such as ( ./folder/games/system/game.zip ) but rather ( ./folder/games/system )

If the folder that relativeTo points to doesn't even exist then we have bigger problems.

**Comment by aarononeal** [comment]
The case you are pointing out doesn’t need to happen in the trusted game list path. Hence it is bypassed.

**Comment by aarononeal** [comment]
Can we get this merged please?

**Comment by joolswills** [comment]
I don't feel it's been explained yet in a way that has convinced us it's wanted as is.

**Comment by aarononeal** [comment]
Please clarify where you are stuck.

1. Touching the file system for thousands of games is extremely slow, particularly over network drives or high latency storage.
2. Trusted game list doesn't require touching the file system.
3. This removes stat calls which touch the file system when using a trusted game list.

File: es-app/src/Gamelist.cpp [hunk 1]
```diff
@@ -165,7 +165,7 @@ void addFileDataNode(pugi::xml_node& parent, const FileData* file, const char* t
 		//there's something useful in there so we'll keep the node, add the path
 
 		// try and make the path relative if we can so things still work if we change the rom folder location in the future
-		newNode.prepend_child("path").text().set(Utils::FileSystem::createRelativePath(file->getPath(), system->getStartPath(), false).c_str());
+		newNode.prepend_child("path").text().set(Utils::FileSystem::createRelativePath(file->getPath(), system->getStartPath(), false, false).c_str());
```
Comment (pjft) [inline es-app/src/Gamelist.cpp:168]: Why change this to "false", though? It has nothing to do with "Trusted gamelists" and this code runs on save. Can you elaborate?

File: es-app/src/Gamelist.cpp [hunk 2]
```diff
@@ -234,7 +234,7 @@ void updateGamelist(SystemData* system)
 					continue;
 				}
 
-				std::string nodePath = Utils::FileSystem::getCanonicalPath(Utils::FileSystem::resolveRelativePath(pathNode.text().get(), system->getStartPath(), true));
+				std::string nodePath = Utils::FileSystem::getCanonicalPath(Utils::FileSystem::resolveRelativePath(pathNode.text().get(), system->getStartPath(), true, false));
```
Comment (pjft) [inline es-app/src/Gamelist.cpp:237]: Same here - this is on save. Why are we removing this check here?

File: es-app/src/MetaData.cpp [hunk 3]
```diff
@@ -112,7 +112,7 @@ void MetaDataList::appendToXML(pugi::xml_node& parent, bool ignoreDefaults, cons
 			// try and make paths relative if we can
 			std::string value = mapIter->second;
 			if (mddIter->type == MD_PATH)
-				value = Utils::FileSystem::createRelativePath(value, relativeTo, true);
+				value = Utils::FileSystem::createRelativePath(value, relativeTo, true, false);
```
Comment (pjft) [inline es-app/src/MetaData.cpp:115]: Same remark.

**Comment by pjft** [comment]
@aarononeal to be clear, I tested it in a network file system with 64 [systems + collections] and 4.800 entries, using "Parse Gamelist Only". I did not not find any meaningful performance difference, as mentioned, but I did concede that I'd like others to test this and report back as there might be other factors here - and I have recently implemented some code to cache some stat calls, but clearly these I missed. 

@tomaz82 , who wrote this code, is also involved, and I think all of us are keeping this conversation going. As @joolswills said, we still need more clarification of what this effectively does.

What you say is absolutely true: touching the file system is a costly operation, trusted gamelist shouldn't require that. It is not in any way clear, however, how these changes directly affect that - and don't bring out other unexpected behaviors.

I managed to understand the isolated path where the stat calls are executed on load - vs on exit, which I left some comments to you as to why we want to change that behavior as well.

Nobody here has anything to gain in not merging a PR that may be helpful, but please understand that we'd like to get everyone to the same page in regards to this specific change, as it is fairly core.

By the way, have you shared this in the forums and have others test this?

Thanks.

**Comment by pjft** [comment]
@aarononeal , building on this, and focusing on the paths that "load" the gamelist, these are the arguments:
- in findOrCreateFile, "system->getRootFolder()" - which'd better be a folder;
- in parseGamelist, "system->getStartPath()" - which, once again, would better be a folder;

They actually don't seem to depend on the gamelist being trusted or not, or am I missing something?

As for the remarks I made for saving gamelists, I would like to know more about the change on that end in order to be fully comfortable with those. Once again, not saying they're incorrect, but would appreciate everybody getting on the same page as this is fairly critical code.

Furthermore, there's a question from @joolswills around symlinks and whether this would break any setups that rely on symlinks. Have you tested such scenarios, and can you elaborate on that front?

Thanks.

File: es-app/src/Gamelist.cpp (hunk 1 above)
Comment (aarononeal) [reply es-app/src/Gamelist.cpp:168]: Because `system->getStartPath()` is always a folder there is no reason to hit the file system in the read or write case.

If there's some concern about it not being a folder, and I don't know what that would be, checking it once up front would be far better than checking it every single file of the game list.

File: es-app/src/Gamelist.cpp (hunk 2 above)
Comment (aarononeal) [reply es-app/src/Gamelist.cpp:237]: Same remark.

File: es-app/src/MetaData.cpp (hunk 3 above)
Comment (aarononeal) [reply es-app/src/MetaData.cpp:115]: Same remark.

**Comment by aarononeal** [comment]
> I did not not find any meaningful performance difference

@pjft Glad you have a low latency network setup. However, on high latency storage clusters the difference is significant. I reported my findings and it is pretty easy to notice the difference between a load time of 8 minutes and one of 3 seconds.

> * in findOrCreateFile, "system->getRootFolder()" - which'd better be a folder;
> * in parseGamelist, "system->getStartPath()" - which, once again, would better be a folder;
>
> They actually don't seem to depend on the gamelist being trusted or not, or am I missing something?

Those are indeed always a folder and there is no logical reason I can see they would ever be a filename path. Therefore, they can always take the false path of traversing the path hierarchy.

> ... whether this would break any setups that rely on symlinks

I have no idea what the concern would be here, and in my tests works fine.

**Comment by pjft** [comment]
@aarononeal Coming back to this, adding two notes.

There have recently been some changes (that I just approved) that may affect the performance in a network environment. Based on results from the #716 Profiling Util, applying #686 and #688 seems to help with these scenarios by a meaningful amount.

Could I ask you to re-test the current master in your setup, and if it's still slower than with your changes, could you run it with #716 merged (you need to compile with the debug flag, and then run ES with the debug command line option as well) to get a proper profiling around the key methods affected - which I believe would be the stat calls as you mentioned?

Thanks.

**Comment by tomaz82** [comment]
Closing this since we never heard back from the author.

**Comment by aarononeal** [comment]
@tomaz82 @pjft I’m using clustered storage with high latency, so the problem is making the stat64 call at all when the game list is trusted. It’s not about reducing the calls, it must eliminate them entirely. I don’t see that the noted PRs address the code paths I reported, so there is no reason to test.

**Comment by tomaz82** [comment]
@aarononeal I made a new PR since your PR needed both redoing the logic of the variable (still has nothing with parents to do) and a rebase.
https://github.com/RetroPie/EmulationStation/pull/780

Is this ok? This should limit the amount of stat64 calls down to a minimum.

**Comment by aarononeal** [comment]
@tomaz82 Yes, I think the revised PR looks good. Thanks for doing that.
